Deep copy schemas and schema in handlers

derekbit opened the pull request:
The shallow copies of the schemas and schema lead to the concurrent map writes fatal error.

Related issue: https://github.com/longhorn/longhorn/issues/3077
Reproduce steps: https://github.com/longhorn/longhorn/issues/3077#issuecomment-1092544440

Signed-off-by: Derek Su <derek.su@suse.com>

derekbit committed:
Add deep copy library

Signed-off-by: Derek Su <derek.su@suse.com>

shuo-wu commented on api/handler.go:
```diff
@@ -88,11 +89,17 @@ func VersionHandler(schemas *client.Schemas, version string) http.Handler {
+			rw.WriteHeader(500)
+			return
+		}
+		schemasCopy := copy.(*client.Schemas)
```
Should we add the assertion check to avoid unexpected results from the deep copy? e.g.,
```
schemasCopy, ok := copy.(*client.Schemas)
if !ok {
	logrus.WithField("err", err).Errorf("Invalid deep copy schema")
	rw.WriteHeader(500)
	return
}
```

derekbit replied:
Good idea! I will update it. Thank you.

derekbit committed:
Deep copy schemas and schema in handlers

The shallow copies of the schemas and schema lead to the concurrent
map writes fatal error.

Signed-off-by: Derek Su <derek.su@suse.com>

shuo-wu commented on api/handler.go:
```diff
@@ -88,11 +89,25 @@ func VersionHandler(schemas *client.Schemas, version string) http.Handler {
+			return
+		}
+
+		schemasCopy, ok := copy.(*client.Schemas)
```
Is this `.(*client.Schemas)` or `.(client.Schemas)`? Similar below.

derekbit replied:
`schemas`'s type is `*client.Schemas`, so use `.(*client.Schemas)`.

derekbit commented on api/handler.go:
```diff
@@ -107,9 +122,22 @@ func SchemaHandler(schemas *client.Schemas) http.Handler {
 		apiContext := GetApiContext(r)
 
 		schema := schemas.Schema(mux.Vars(r)["id"])
+		copy, err := copystructure.Copy(schema)
+		if err != nil {
+			logrus.WithField("err", err).Errorf("Failed to deep copy schema")
+			rw.WriteHeader(500)
+			return
+		}
+
+		schemaCopy, ok := copy.(client.Schema)
```
`schema`'s type is `client.Schema`, so use `.(client.Schema)` here.
